lucene/core/src/java/org/apache/lucene/search/ParallelSortedCollector.java

```diff
@@ -0,0 +1,612 @@
+      return numHits > TERMINATION_STRATEGY_HIT_THRESHOLD && sortIsNumeric(sort);
+    }
+
+    private static boolean sortIsNumeric(Sort sort) {
+      switch (sort.getSort()[0].getType()) {
+        case DOC:
+        case INT:
+        case FLOAT:
+        case LONG:
+        case DOUBLE:
+          return true;
+        default:
+          return false;
+      }
+    }
+
+    synchronized LeafState addLeafState() {
+      LeafState newLeafState = new LeafState();
+      newLeafState.index = leafStates.size();
+      leafStates.add(newLeafState);
+      scratch.add(new LeafState());
+      return newLeafState;
+    }
+
+    /**
+     * Set the update interval to 2^bitCount and the intervalMask to 2^bitCount-1. This controls the
+     * rate at which multiple threads report their worst scores. For best performance this should be
+     * set to the nearest power of 2 &gt; the number of expected calling threads.
+     * @param bitCount the number of bits in the interval/intervalMask
+     */
+    void setIntervalBits(int bitCount) {
+      interval = 1 << bitCount;
+      intervalMask = interval - 1;
+    }
+
+    /**
+     * Must be called by leaf collectors on every interval hits to update their progress; they should
+     * call when ((leafCollector.numCollected &amp; this.intervalMask) == 0).
+     * @param newLeafState the leaf collector's current lowest score
+     * @return whether the collector should terminate
+     */
+    synchronized boolean updateLeafState(LeafState newLeafState) {
+      if (newLeafState.isActive()) {
+        // Only count hits from this leaf if it is still active: otherwise its hits are not competitive and do not count
+        // towards meeting the threshold for terminating the remaining leaves
+        totalCollected += interval;
+      }
+      // (1) Until we collect totalToCollect we can't terminate anything.
+      // (2) after that, any single leaf that has collected numHits can be terminated,
+      // (3) and we may be able to remove the worst leaves even if they have not yet collected numHits, as long as the
+      // remaining leaves have collected totalToCollect.
+      if (totalCollected >= totalToCollect) {
+        if (newLeafState.resultCount >= numHits) {
+          // This leaf has collected enough hits all on its own
+          return true;
+        }
+        // We may be ready to terminate some leaves (although they may not include the one that just called
+        // updateLeafState(), such leaves will eventually be terminated here).
+        excludeSuperfluousLeaves();
+        if (newLeafState.compareTo(thresholdState) <= 0) {
+          // Tell the current leaf collector to terminate since it can no longer contribute any top hits
+          //System.out.println("  terminate " + newLeafState + " b/c it is > " + thresholdState);
+          return true;
+        }
+      }
+      return false;
+    }
+
+    private void excludeSuperfluousLeaves() {
```

Comment:
This whole logic seems wrong to me. We should track state on the leaves directly.

Reply:
I tried to explain above -- this is the whole idea, really. Maybe there is something else we can do that would be as good or better, but I don't see it yet. Given the gains here, I'm tempted to go forward and then I'd be *perfectly happy* to see this all thrown away in favor of something neater if we can achieve the same result.